main/src/ca/uwaterloo/flix/language/phase/unification/BoolFormulaSubstitution.scala

```diff
@@ -0,0 +1,135 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.ast.Symbol
+import ca.uwaterloo.flix.util.InternalCompilerException
+import ca.uwaterloo.flix.util.collection.Bimap
+
+/**
+  * Companion object for the [[BoolFormulaSubstitution]] class.
+  */
+object BoolFormulaSubstitution {
+  /**
+    * Returns the empty substitution.
+    */
+  def empty[F]: BoolFormulaSubstitution[F] = BoolFormulaSubstitution(Map.empty)
+
+  /**
+    * Returns the singleton substitution mapping the type variable `x` to `tpe`.
+    */
+  def singleton[F](x: Int, f: F): BoolFormulaSubstitution[F] = {
+    // Ensure that we do not add any x -> x mappings.
+    f match {
+      case y: ExplicitFormula.Var if x == y.x => empty
+      case _ => BoolFormulaSubstitution(Map(x -> f))
+    }
+  }
+
+}
+
+/**
+  * A substitution is a map from type variables to types.
+  */
+case class BoolFormulaSubstitution[F](m: Map[Int, F]) {
+
+  /**
+    * Returns `true` if `this` is the empty substitution.
+    */
+  val isEmpty: Boolean = m.isEmpty
+
+  /**
+    * Converts this formula substitution into a type substitution
+    */
+  def toTypeSubstitution(env: Bimap[Symbol.KindedTypeVarSym, Int])(implicit alg: BoolFormula[F]): Substitution = {
+    val map = m.map {
+      case (k0, v0) =>
+        val k = env.getBackward(k0).getOrElse(throw InternalCompilerException(s"missing key $k0"))
+        val v = alg.toType(v0, env)
+        (k, v)
+    }
+    Substitution(map)
+  }
+
+  /**
+    * Applies `this` substitution to the given type `tpe0`.
+    */
+  def apply(f: F)(implicit alg: BoolFormula[F]): F = {
+    // Optimization: Return the type if the substitution is empty. Otherwise visit the type.
+    if (isEmpty) {
+      f
+    } else {
+      alg.map(f) {
+        i => m.getOrElse(i, alg.mkVar(i))
+      }
+    }
+  }
+
+  /**
+    * Applies `this` substitution to the given types `ts`.
+    */
+  def apply(ts: List[F])(implicit alg: BoolFormula[F]): List[F] = if (isEmpty) ts else ts map apply
+
+  /**
+    * Returns the left-biased composition of `this` substitution with `that` substitution.
+    */
+  def ++(that: BoolFormulaSubstitution[F]): BoolFormulaSubstitution[F] = {
+    if (this.isEmpty) {
+      that
+    } else if (that.isEmpty) {
+      this
+    } else {
+      BoolFormulaSubstitution(
+        this.m ++ that.m.filter(kv => !this.m.contains(kv._1))
+      )
+    }
+  }
+
+  /**
+    * Returns the composition of `this` substitution with `that` substitution.
+    */
+  def @@(that: BoolFormulaSubstitution[F])(implicit alg: BoolFormula[F]): BoolFormulaSubstitution[F] = {
```

Comment:
since there is no coherence, should the implicit boolformula actually be an implicit of the environment and not of the function?

Reply:
I guess that does not work with empty.. OK.